### DISCUSSION TITLE
Get currently playing song (not playing in librespot)

### DISCUSSION THREAD

**Discussion by BenJeau** [librespot-org/librespot#1028] [General]
Hi, is there a way to know what is being currently played, if the song is being played from another source (ex: from the official Spotify client)?

Is there something like that in the lib?

As a backup, I can get a token (like here https://github.com/librespot-org/librespot/blob/dev/examples/get_token.rs) and then periodically query https://developer.spotify.com/console/get-users-currently-playing-track/, but since I'd like for it to be near real-time, I'll query it almost every second (and unsure if there's a rate limit to Spotify's API) - so I'm curious if there's a better way to go about this.

Thanks!

**Comment by roderickvd** [comment]
No, we currently do not expose that or even deal with those messages. Although you could if someone added this. Not trivial but not super hard either.

**Reply by BenJeau** [reply]
Thanks for the quick response! I don't mind diving in the code, I'm looking more in the codebase and I'm seeing a bunch of protobufs, I'm guessing if there's a message being sent it would be via protobuf? I'm trying to dissect the traffic via wireshark (although I'm a noob when it comes to wireshark and I'm getting no where with that + I think the traffic is probably encrypted).

I see some usage of the protocol crate in core/connect/metadata crates, is there a place in the code where you'd recommend for me to look at as an example of pulling/listening to those messages?

**Reply by BenJeau** [reply]
And is there a way you'd go about to reverse engineer the messages (or it's pretty straightforward because of the protobufs)?

**Reply by roderickvd** [reply]
I am no expert in intercepting on the wire either. I just dump what’s received over Mercury or HTTP. You can find both in `core`. 

Usually because you know the endpoint (or have intercepted it — that’s easier certainly for HTTP) then you can guess the name of the protobuf. 

There is a tool in Google’s toolbox — can’t remember which now — that prints the protobuf format when you feed it a binary protobuf. The fields are then anonymous of course but it can help finding out the right protobuf struct. 

Everyone needs to start somewhere or course but honestly I don’t know if reverse engineering Spotify would be the easiest…
